shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

```diff
@@ -182,6 +183,117 @@
   return canvases;
 }
 
+int FlutterPlatformViewsController::GetNumberOfClips(const MutatorsStack& mutators_stack) {
+  // Apply transforms/clips.
+  std::vector<std::shared_ptr<Mutator>>::const_reverse_iterator iter = mutators_stack.bottom();
+  int clipCount = 0;
+  while (iter != mutators_stack.top()) {
+    if ((*iter)->isClipType()) {
+      clipCount++;
+    }
+    ++iter;
+  }
+  return clipCount;
+}
+
+UIView* FlutterPlatformViewsController::ReconstructClipViewsChain(int number_of_clips,
+                                                                  UIView* child,
+                                                                  UIView* parent) {
+  NSInteger index = -1;
+  if (parent.superview) {
+    // TODO(cyanglaz): protentially cache the index of oldPlatformViewRoot to make this a O(1).
+    index = [flutter_view_.get().subviews indexOfObject:parent];
+    [parent removeFromSuperview];
+  }
+  UIView* head = child;
+  BOOL needAddNewView = NO;
+  for (int i = 0; i < number_of_clips; i++) {
+    if (head == parent) {
+      needAddNewView = YES;
+    }
+    if (needAddNewView) {
+      ChildClippingView* clippingView = [ChildClippingView new];
+      [clippingView addSubview:head];
+      head = clippingView;
+    } else {
+      head = head.superview;
+    }
+  }
+  if (!needAddNewView && head != parent) {
+    [head removeFromSuperview];
+  }
+  if (index > -1) {
+    [flutter_view_.get() insertSubview:head atIndex:index];
+  }
+  return head;
+}
+
+void FlutterPlatformViewsController::ApplyMutators(const MutatorsStack& mutators_stack,
+                                                   UIView* embedded_view) {
+  UIView* head = embedded_view;
+  head.clipsToBounds = YES;
+  head.layer.transform = CATransform3DIdentity;
+  ResetAnchor(head.layer);
+
+  std::vector<std::shared_ptr<Mutator>>::const_reverse_iterator iter = mutators_stack.bottom();
+  while (iter != mutators_stack.top()) {
+    switch ((*iter)->type()) {
+      case transform: {
+        CATransform3D transform = GetCATransform3DFromSkMatrix((*iter)->matrix());
+        head.layer.transform = CATransform3DConcat(head.layer.transform, transform);
+        break;
+      }
+      case clip_rect:
+      case clip_rrect:
+      case clip_path: {
+        UIView* clipView = head.superview;
```

Comment:
nit: maybe cast this to a ChildClippingView already here?
We can add a comment saying that this is called after `ReconstructClipViewsChain` which assures we have a ChildClippingView parent for each clip in the stack.